kernel/constr.ml

```diff
@@ -478,6 +335,158 @@ let decompose_app c =
     | App (f,cl) -> (f, cl)
     | _ -> (c,[||])
 
+(*********************)
+(* Term constructors *)
+(*********************)
+
+(* Constructs a de Bruijn index with number n *)
+let rels = Array.init 17 (fun i -> T (Rel i))
+
+let mkRel n = if 0<=n && n<=16 then rels.(n) else T (Rel n)
+
+(* If lt = [t1; ...; tn], constructs the application (t1 ... tn) *)
+(* We ensure applicative terms have at least one argument and the
+   function is not itself an applicative term *)
+let mkApp (f, a) =
+  if Int.equal (Array.length a) 0 then f else
+    match kind f with
+      | App (g, cl) -> T (App (g, Array.append cl a))
+      | _ -> T (App (f, a))
+
+let mkSProp = T (Sort Sorts.sprop)
+let mkProp  = T (Sort Sorts.prop)
+let mkSet   = T (Sort Sorts.set)
+
+(* Enforces:
+   - applicative terms have at least one argument and the
+     function is not itself an applicative term
+   - stacks of VM or native casts are collapsed
+   - small rels are shared
+   - small sorts are shared
+*)
+let of_kind = function
+| Rel n when 0 <= n && n < Array.length rels -> rels.(n)
+| App (f, [||]) -> f
+| App (f, a) when isApp f -> mkApp (f, a)
```

Comment:
I really don't like when clauses, especially in a piece of code so low-level. Couldn't you just write
```ocaml
| App (App _ as f, a) -> mkApp (f, a)
```
or even just inlining mkApp in the branch?

Reply:
ok